src/python/pants/engine/README.md

```diff
@@ -122,7 +122,7 @@ print(x.content)                # 'a string'
 # datatype objects can be easily inspected:
 print(x)                        # 'FormattedInt(content="a string")'
 
-# Most mypy types, including parameterized types, may be used in field definitions.
+# All mypy types, including parameterized types, may be used in field definitions.
```

Comment:
First, I'm very happy with the laziness... Why fetch a file unless you need it. Imagine a large .jar or something...

But then can the second process execution (that takes the Foo.java as input) just reference it via the CAS, so we never even see the full file locally? I.e., the first remote process execution creates it in the CAS, the second reads it from there, and we never need to download the intermediate file to the Pants-running machine, just its digest?

And my other question was more about this: in the remote execution of the proto compiler, something has to know to take the resulting Foo.scala file from the local filesystem and stick it in the CAS (regardless of whether we then fetch it eagerly or lazily). How does that decision get made? I.e., what knows that "Foo.scala" is a file we care about, a product, but "incidental log file.txt" is not...

Reply:
> But then can the second process execution (that takes the Foo.java as input) just reference it via the CAS, so we never even see the full file locally? I.e., the first remote process execution creates it in the CAS, the second reads it from there, and we never need to download the intermediate file to the Pants-running machine, just its digest?

Correct. But whether you download it or not (ie, lazy or no), it already exists in the remote CAS. Before executing the next process we do (confirmed) eagerly call https://github.com/pantsbuild/pants/blob/70cb3c3598d9ec79eb21d1b3c528bd8ee1b82c36/src/rust/engine/fs/store/src/lib.rs#L394-L400: if the inputs already exist remotely, this is a noop.

> And my other question was more about this: in the remote execution of the proto compiler, something has to know to take the resulting Foo.scala file from the local filesystem and stick it in the CAS (regardless of whether we then fetch it eagerly or lazily). How does that decision get made? I.e., what knows that "Foo.scala" is a file we care about, a product, but "incidental log file.txt" is not...

The high-level (python) `ExecuteProcessRequest` type has fields to define what should be captured from the process-execution sandbox, and these map ~1:1 with the bazel remoting API: https://github.com/pantsbuild/pants/blob/e3cf637cc72ba10f34bd6eb2d11ac6242e819b72/src/python/pants/engine/isolated_process.py#L21-L22

Reply:
OK so, to sum up:

Say we're running Pants on machine P1, and executing processes remotely on machines M1, M2.

1. `Foo.proto` is a local source file, so it's already in the CAS and P1 has its fingerprint.

2. P1 calls `ExecuteProcess(protoc, Fingerprint(Foo.proto))`, sending the request to M1.

3.  M1 fetches `Foo.proto` from the CAS, runs `protoc` on it (I'm eliding the question of where M1 gets a `protoc` binary...) thus generating `Foo.scala`. 

4. M1 puts `Foo.scala` in the CAS, returning just `Fingerprint(Foo.scala)` to P1.

5. P1 calls `ExecuteProcess(scalac, Fingerprint(Foo.scala))`, sending the request to M2.

6. M2 fetches `Foo.scala` from the CAS, runs `scalac` on it, thus generating `Foo.class`.

7. M2 puts `Foo.class` in the CAS, returning just `Fingerprint(Foo.class)` to P1.

So P1 does not have the content of `Foo.scala` or `Foo.class` locally at any point. 

Does that about sum it up reasonably?

If so, one more question: what does that mean for running `ensure_remote_has_recursive` before step 5 above? P1 doesn't have `Foo.scala`'s content, so it seems like it can't ensure anything. It can I guess error out if the CAS is missing the file, but then so can M2 directly.